google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -660,10 +691,72 @@ public Void apply(BulkWriterOperation operation) {
     synchronized (lock) {
       verifyNotClosedLocked();
       writesEnqueued = true;
-      sendOperationLocked(enqueueOperationOnBatchCallback, operation);
+
+      // Advance the lastOperation pointer. This ensures that lastOperation only completes when
+      // both the previous and the current write complete.
+      lastOperation =
+          ApiFutures.transformAsync(
+              lastOperation,
+              new ApiAsyncFunction<Void, Void>() {
+                @Override
+                public ApiFuture<Void> apply(Void aVoid) {
+                  return silenceFuture(operation.getFuture());
+                }
+              },
+              MoreExecutors.directExecutor());
+
+      // Schedule the operation if the BulkWriter has fewer than the maximum number of allowed
+      // pending operations, or add the operation to the buffer.
+      if (pendingOpsCount < maxPendingOpCount) {
+        pendingOpsCount++;
+        sendOperationLocked(enqueueOperationOnBatchCallback, operation);
+      } else {
+        bufferedOperations.add(
+            new Runnable() {
+              @Override
+              public void run() {
+                synchronized (lock) {
+                  pendingOpsCount++;
+                  sendOperationLocked(enqueueOperationOnBatchCallback, operation);
+                }
+              }
+            });
+      }
     }
 
-    return operation.getFuture();
+    ApiFuture<WriteResult> transformedFuture =
+        ApiFutures.transformAsync(
+            operation.getFuture(),
+            new ApiAsyncFunction<WriteResult, WriteResult>() {
+              public ApiFuture<WriteResult> apply(WriteResult result) throws Exception {
+                pendingOpsCount--;
+                processBufferedOperations();
+                return ApiFutures.immediateFuture(result);
+              }
+            });
```

Comment:
Does this not need an executor for the non-deprecated version?

Reply:
added executor, thanks for catching!